src/KDMqtt/mqtt.cpp

```diff
@@ -0,0 +1,523 @@
+        return MOSQ_ERR_UNKNOWN;
+    }
+
+    std::optional<const std::string> nullopt;
+    const auto result = m_mosquitto.client()->tlsSet(cafile.path(), nullopt, nullopt, nullopt);
+    MqttLib::instance().checkMosquittoResultAndDoDebugPrints(result, "MqttClient::setTls()");
+
+    return result;
+}
+
+int MqttClient::setUsernameAndPassword(const std::string &username, const std::string &password)
+{
+    spdlog::debug("MqttClient::setUsernameAndPassword()");
+
+    if (connectionState.get() != ConnectionState::DISCONNECTED) {
+        spdlog::error("MqttClient::setUsernameAndPassword() - Setting AUTH is only allowed when disconnected.");
+        return MOSQ_ERR_UNKNOWN;
+    }
+
+    const auto result = m_mosquitto.client()->usernamePasswordSet(username, password);
+    MqttLib::instance().checkMosquittoResultAndDoDebugPrints(result, "MqttClient::setUsernameAndPassword()");
+    return result;
+}
+
+int MqttClient::setWill(const std::string &topic, int payloadlen, const void *payload, int qos, bool retain)
+{
+    spdlog::debug("MqttClient::setWill() - topic:{}, qos:{}, retain:{})", topic, qos, retain);
+
+    if (connectionState.get() != ConnectionState::DISCONNECTED) {
+        spdlog::error("MqttClient::setWill() - Setting will is only allowed when disconnected.");
+        return MOSQ_ERR_UNKNOWN;
+    }
+
+    const auto result = m_mosquitto.client()->willSet(topic.c_str(), payloadlen, payload, qos, retain);
+    MqttLib::instance().checkMosquittoResultAndDoDebugPrints(result, "MqttClient::setWill()");
+    return result;
+}
+
+int MqttClient::connect(const Url &host, int port, int keepalive)
+{
+    spdlog::debug("MqttClient::connect() - host:{}, port:{}, keepalive:{})", host.url(), port, keepalive);
+
+    if (connectionState.get() == ConnectionState::CONNECTING) {
+        spdlog::error("MqttClient::connect() - Already connecting to host.");
+        return MOSQ_ERR_UNKNOWN;
+    }
+
+    if (connectionState.get() == ConnectionState::CONNECTED) {
+        spdlog::error("MqttClient::connect() - Already connected to a host. Disconnect from current host first.");
+        return MOSQ_ERR_UNKNOWN;
+    }
+
+    connectionState.set(ConnectionState::CONNECTING);
+    // TODO -> TLS only works with connect(), it does not work with connect_async()
+    // I'm uncertain if there is a setup that allows for connect_async() with TLS (I didn't manage to find one)
+    // (the use of non-blocking connect_async() would be preferred from our POV)
+    // other people seem to have encountered similiar behaviour before, though this issue should have been fixed a while ago
+    // -> https://github.com/eclipse/mosquitto/issues/990
+    const auto start = clock();
+    const auto result = m_mosquitto.client()->connect(host.url().c_str(), port, keepalive);
+    const auto end = clock();
+    const auto elapsedTimeMs = std::round((double(end - start) / double(CLOCKS_PER_SEC)) * 1000000.0);
```

Comment:
I'd rather do this with chrono `high_resolution_clock`. Haven't checked but it should be something like:
```suggestion
    {
        using namespace std::chrono;
        const auto start = high_resolution_clock::now();
        const auto result = m_mosquitto.client()->connect(host.url().c_str(), port, keepalive);
        const auto end = high_resolution_clock::now();
        const duration elapsed = duration_cast<milliseconds>(end - start);
        spdlog::info("MqttClient::connect() - blocking call of MosquittoClient::connect() took {} µs", elapsed.count());
    }
```

Reply:
I found a way to use `mosquitto_connect_async` which also makes most of this code block obsolete.
I pushed the change to use async connect as a new commit. Let me know what you think about it.